File: RecordDatabasePython/tests/test_dialogs.py
import sys
import unittest
from PyQt5 import QtTest

from PyQt5.QtWidgets import QApplication
from PyQt5.QtTest import QTest

from recorddatabase.dialogs import NewRecordDialog

app = QApplication(sys.argv)

class TestNewRecordDialog(unittest.TestCase):
    def setUp(self):
        self.dialog = NewRecordDialog()

    def test_construction(self):
        self.assertIsNotNone(self.dialog.idLineEdit, "LineEdit ID not created")
        self.assertIsNotNone(self.dialog.nameLineEdit, "LineEdit name not created")
        self.assertIsNotNone(self.dialog.surnameLineEdit, "LineEdit surname not created")

        self.assertIsNotNone(self.dialog.createRecordButton, "Create record button not created")

    def test_fields(self):
        validId = "123123"
        invalidId1 = "123"
        invalidId2 = "123asd"

        QTest.keyClicks(self.dialog.idLineEdit, validId)
        self.assertEqual(self.dialog.idLineEdit.text(), validId)

        self.dialog.idLineEdit.clear()
        QTest.keyClicks(self.dialog.idLineEdit, invalidId1)
        self.assertEqual(self.dialog.idLineEdit.text(), invalidId1)

        self.dialog.idLineEdit.clear()
        QTest.keyClicks(self.dialog.idLineEdit, invalidId2)
        self.assertEqual(self.dialog.idLineEdit.text(), "123")

    def test_acceptButton(self):
        validId = "123123"
        invalidId1 = "123"

        self.assertFalse(self.dialog.createRecordButton.isEnabled(), "Button is enabled at beginning")

        QTest.keyClicks(self.dialog.idLineEdit, invalidId1)
        self.assertFalse(self.dialog.createRecordButton.isEnabled(), "Button is enabled after invalid input")

        self.dialog.idLineEdit.clear()
        QTest.keyClicks(self.dialog.idLineEdit, validId)
        self.assertTrue(self.dialog.createRecordButton.isEnabled(), "Button is enabled after valid input")


if __name__ == "__main__":
    unittest.main()

File: RecordDatabaseNative/test/test_dialogs.cpp
#include <QtTest>
#include <QCoreApplication>
#include <QtWidgets>

#include "dialogs.h"

class TestNewRecordDialog : public QObject
{
    Q_OBJECT

public:
    TestNewRecordDialog();
    ~TestNewRecordDialog();

private slots:
    void initTestCase();  // before first test case
    void cleanupTestCase();  // after last test case

    void init();  // before each case
    void cleanup();  // after each case

    void TestConstruction();
    void TestFields();
    void TestAcceptButton();

private:
    NewRecordDialog newRecordDialog;
};

TestNewRecordDialog::TestNewRecordDialog()
{

}

TestNewRecordDialog::~TestNewRecordDialog()
{

}

void TestNewRecordDialog::initTestCase()
{
    qDebug("Called before everything else.");
}

void TestNewRecordDialog::cleanupTestCase()
{
    qDebug("Called after tests run.");
}

void TestNewRecordDialog::init()
{
    qDebug("Called before each case.");
    newRecordDialog.open();
}

void TestNewRecordDialog::cleanup()
{
    qDebug("Called after each case.");
    newRecordDialog.close();
}

void TestNewRecordDialog::TestConstruction()
{
    QVERIFY2(newRecordDialog.m_idLineEdit, "LineEdit ID not created");
    QVERIFY2(newRecordDialog.m_nameLineEdit, "LineEdit name not created");
    QVERIFY2(newRecordDialog.m_surnameLineEdit, "LineEdit surname not created");

    QVERIFY2(newRecordDialog.m_createRecordButton, "Create record button not created");
}

void TestNewRecordDialog::TestFields()
{
    QString validId("123123");
    QString invalidId1("123");
    QString invalidId2("123asd");

    QTest::keyClicks(newRecordDialog.m_idLineEdit, validId);
    QCOMPARE(newRecordDialog.m_idLineEdit->text(), validId);

    newRecordDialog.m_idLineEdit->clear();
    QTest::keyClicks(newRecordDialog.m_idLineEdit, invalidId1);
    QCOMPARE(newRecordDialog.m_idLineEdit->text(), invalidId1);

    newRecordDialog.m_idLineEdit->clear();
    QTest::keyClicks(newRecordDialog.m_idLineEdit, invalidId2);
    QCOMPARE(newRecordDialog.m_idLineEdit->text(), "123");
}

void TestNewRecordDialog::TestAcceptButton()
{
    QString validId("123123");
    QString invalidId1("123");

    QVERIFY2(!newRecordDialog.m_createRecordButton->isEnabled(), "Button is enabled at beginning");

    QTest::keyClicks(newRecordDialog.m_idLineEdit, invalidId1);
    QVERIFY2(!newRecordDialog.m_createRecordButton->isEnabled(), "Button is enabled after invalid input");

    newRecordDialog.m_idLineEdit->clear();
    QTest::keyClicks(newRecordDialog.m_idLineEdit, validId);
    QVERIFY2(newRecordDialog.m_createRecordButton->isEnabled(), "Button is disabled after valid input");
}

QTEST_MAIN(TestNewRecordDialog)

#include "test_dialogs.moc"
